File: Python/watertank_n_tank.py
import collections
import sys

# Getting parameters
vol_wanted = int(sys.argv[1])
containers_count = int(sys.argv[2])
containers_volume = []
for i in range(0, containers_count):
    containers_volume.append(int(sys.argv[i+3]))
check_dict = {}
starting_node = [[]]
for i in range(0, containers_count):
    starting_node[0].append(0)


# def check_args(cont_vol, vol_wanted):
#     """
#         If water required is superior of the biggest tank, there's no solution
#         With the Bezout Lemma we can check if there's a solution such as + by = z
#         Lemma : if x and y are nonzero integers and g = gcd(x,y),
#         then there exist integers a and b such that ax+by=g.
#     """
#     if vol_wanted > max(cont_vol[0], cont_vol[1]) or (vol_wanted % (gcd(cont_vol[0], cont_vol[1])) is not 0):
#         print("No solution")
#         return False
#     return True


def search(starting_node, containers_volume, vol_wanted, check_dict):
    # Starting state
    target = []
    accomplished = False

    # Using collections and deque to have a list with improved method appendleft, popleft etc..
    q = collections.deque()
    q.appendleft(starting_node)

    # if check_args(containers_volume, vol_wanted):
    while len(q) != 0:
        path = q.popleft()
        check_dict[get_index(path[-1])] = True

        # If target is reached, exit the loop
        if is_vol_wanted(path, vol_wanted):
            accomplished = True
            target = path
            break

        # If target no reached, getting the next move
        next_moves = next_step(containers_volume, path, check_dict)
        for i in next_moves:
            q.append(i)

    # If accomplished, print the path
    if accomplished:
        print_path(target)
    else:
        print("No Solution")
        return


def been_there(node, check_dict):
    # Check if a node had already been visited

    return check_dict.get(get_index(node), False)


def next_step(containers_volume, path, check_dict):
    # Finding the next path with checking the ones already checked

    result = []
    next_nodes = []
    max_containers_vol = []
    containers_status = []
    the_node =[]

    for i in range(0, containers_count):
        max_containers_vol.append(containers_volume[i])

    for i in range(0, containers_count):
        containers_status.append(path[-1][i])

    # Filling tanks one by one and add the path
    for j in range(0, containers_count):
        for i in range(0, containers_count):
            the_node.append(containers_status[i])
        the_node[j] = max_containers_vol[j]
        if not been_there(the_node, check_dict):
            next_nodes.append(the_node)
        the_node = []

    # Empty tank one by one and add the path
    for j in range(0, containers_count):
        for i in range(0, containers_count):
            the_node.append(containers_status[i])
        the_node[j] = 0
        if not been_there(the_node, check_dict):
            next_nodes.append(the_node)
        the_node = []

    for i in range(0, containers_count):
        for j in range(0, containers_count):
            if j != i:
                new_node = containers_status[:]
                new_node[i] = (min(max_containers_vol[i], new_node[i] + new_node[j]))
                new_node[j] = (new_node[j] - (new_node[i] - containers_status[i]))
                if not been_there(new_node, check_dict):
                    next_nodes.append(new_node)

    # Create the list of possible next path
    for i in range(0, len(next_nodes)):
        temp = list(path)
        temp.append(next_nodes[i])
        result.append(temp)

    return result


def write_transition(old, new):
    alphabet = ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N"]
    diff_count = []
    for index, (first, second) in enumerate(zip(old, new)):
        if first != second:
            diff_count.append(index)
    if len(diff_count) == 1:
        if new[diff_count[0]] == 0:
            return f'{alphabet[diff_count[0]]}->*'
        else:
            return f'*->{alphabet[diff_count[0]]}'
    else:
        if new[diff_count[0]] > old[diff_count[0]]:
            return f'{alphabet[diff_count[1]]}->{alphabet[diff_count[0]]}'
        else:
            return f'{alphabet[diff_count[0]]}->{alphabet[diff_count[1]]}'


def print_path(path):
    print(f'Solution in {len(path) - 1} steps')
    for i in range(0, len(path) - 1):
        print(write_transition(path[i], path[i + 1]), tuple(path[i + 1]))


def is_vol_wanted(path, vol_wanted):
    # Checking the last node and return True if we reached the targeted value
    for i in range(0, containers_count):
        if path[-1][i] == vol_wanted:
            return True
    return False


def get_index(node):
    # generating a random number for the key index using " ** " to be sure there's no duplicate
    result_ind = 1
    for i in range(0, containers_count):
        result_ind *= (3+i) ** node[i]
    return result_ind


(search(starting_node,containers_volume,vol_wanted,check_dict))

File: Java/src/Watertank_n_tank.java
import java.util.ArrayList;
import java.util.HashMap;

// dynamic programming method using Breadth-First Search (BFS) Algorithm

public class Watertank_n_tank {

    public static void main(String[] args) {
        // General Parameters
        ArrayList<Integer> containers_volume = new ArrayList<>();
        ArrayList<Integer> starting_node = new ArrayList<>();

        int vol_wanted=Integer.parseInt(args[0]);
        int containers_count = Integer.parseInt(args[1]);

        for(int i =0; i<containers_count; i++) {
            starting_node.add(0);
            containers_volume.add(Integer.parseInt((args[i + 2])));
        }
        HashMap<Double, Boolean>  check_dict = new HashMap<>();



        search(starting_node, containers_volume, vol_wanted, check_dict,containers_count);

    }

    public static void search(ArrayList<Integer> starting_node, ArrayList<Integer> containers_volume, int vol_wanted, HashMap<Double, Boolean>  check_dict, int containers_count)
    {
        ArrayList<ArrayList<Integer>> target=new ArrayList<>();
        ArrayList<ArrayList<Integer>> q =new ArrayList<>();
        ArrayList<ArrayList<Integer>> path = new ArrayList<>();
        ArrayList<ArrayList<Integer>> next_moves;
        boolean accomplished= false;

        q.add(starting_node);


        while(q.size() != 0) {
            ArrayList<Integer> path_2;
            path_2 = q.get(0);
            ArrayList<Integer> comp = new ArrayList<>();
            for(int i =0; i<containers_count; i++)
                comp.add(0);
            if (path_2.equals(comp))
            {
                path.add(new ArrayList<>(path_2));
            }
            else{
                path = (new ArrayList (path_2));
            }
            q.remove(0);

            ArrayList<Integer> last_path;
            last_path = path.get(path.size() - 1);
            double b = get_index(last_path, containers_count);
            check_dict.put(b, true);

            // If target is reached, exit the loop
            if (is_vol_wanted(path, vol_wanted,containers_count)) {
                accomplished = true;
                target = path;
                break;
            }

            // If target no reached, getting the next move
            next_moves = next_step(containers_volume, path, check_dict, containers_count);
            q.addAll(next_moves);
        }
        // If accomplished, print the path
        if(accomplished)
        {
            print_path(target);
            System.out.println(target);
        }
        else
        {
            System.out.println("no solution");
        }

    }

    public static boolean been_there(ArrayList<Integer> node, HashMap<Double, Boolean> check_dict,int containers_count)
    // Check if a node had already been visited
    {
        double a=get_index(node, containers_count);
        return !check_dict.containsKey(a);
    }

    public static ArrayList next_step(ArrayList<Integer> containers_volume, ArrayList<ArrayList<Integer>> path, HashMap<Double, Boolean>  check_dict, int containers_count)
    // Finding the next path with checking the ones already checked
    {
        ArrayList result=new ArrayList<>();
        ArrayList<ArrayList<Integer>> next_nodes=new ArrayList<>();
        ArrayList<Integer> node=new ArrayList<>();
        ArrayList<Integer> last_path;
        ArrayList<Integer> max_containers_volume = new ArrayList<>();
        ArrayList<Integer> containers_status = new ArrayList<>();
        ArrayList<Integer> the_node = new ArrayList<>();

        for(int i =0; i<containers_count; i++)
            max_containers_volume.add(containers_volume.get(i));


        last_path = path.get(path.size()-1);
        for(int i =0; i<containers_count; i++)
            containers_status.add(last_path.get(i));


        // fill in the first tank.
        for(int j =0; j < containers_count; j++) {
            for (int i = 0; i < containers_count; i++)
                the_node.add(containers_status.get(i));
            the_node.set(j, max_containers_volume.get(j));
            if(been_there(the_node, check_dict, containers_count))
                next_nodes.add(new ArrayList<>(the_node));
            the_node.clear();
        }

        for(int j =0; j < containers_count; j++) {
            for (int i = 0; i < containers_count; i++)
                the_node.add(containers_status.get(i));
            the_node.set(j,0);
            if(been_there(the_node, check_dict, containers_count))
                next_nodes.add(new ArrayList<>(the_node));
            the_node.clear();
        }

        // Transfer from tank to another
        for(int i =0; i < containers_count; i++) {
            for (int j = 0; j < containers_count; j++) {
                if (j != i) {
                    the_node = new ArrayList<>(containers_status);
                    the_node.set(i, Math.min(max_containers_volume.get(i), the_node.get(i) + the_node.get(j)));
                    the_node.set(j, (the_node.get(j) - (the_node.get(i) - containers_status.get(i))));
                    if (been_there(the_node, check_dict, containers_count))
                        next_nodes.add(new ArrayList<>(the_node));
                    the_node.clear();
                }
            }

        }

            // Create the list of possible next path
        for (ArrayList arrayList : next_nodes) {
            ArrayList<ArrayList<Integer>> temp;
            temp = new ArrayList<>(path);
            temp.add(arrayList);
            result.add(new ArrayList(temp));
            temp.removeAll(result);
            }

         return result;

    }

    public static void write_transition(ArrayList<Integer> old, ArrayList<Integer> m_new)
    {
        int first_cont= old.get(0);
        int second_cont= old.get(1);
        int new_first_cont= m_new.get(0);
        int new_second_cont= m_new.get(1);

        if(first_cont>new_first_cont){
            if(second_cont==new_second_cont)
            {
                System.out.print("A->* : ");
            }
            else
            {
                System.out.print("A->B : ");
            }

        }
        else
        {
            if (second_cont > new_second_cont) {
                if (first_cont == new_first_cont){
                    System.out.print("B->* : ");
                }
                else
                {
                    System.out.print("B->A : ");
                }
            }
            else
            {
                if(first_cont == new_first_cont)
                {
                    System.out.print("*->B : ");
                }
                else
                {
                    System.out.print("*->A : ");
                }
            }
        }

    }

    public static  void print_path(ArrayList<ArrayList<Integer>> path)
    {
        System.out.println("Solution in "+(path.size()-1)+" steps");
//        for(int i =0; i<path.size()-1; i++)
//        {
//            write_transition(path.get(i), path.get(i+1));
//            String step;
//            step = path.get(i+1).toString().replaceAll("\\[","\\(").replaceAll("]","\\)");
//            System.out.println(step);
//        }
    }

    public static boolean is_vol_wanted(ArrayList<ArrayList<Integer>> path, int vol_wanted, int containers_count)
    {
        //Checking the last node and return True if we reached the targeted value

        int temp4;
        ArrayList<Integer> last_path;
        last_path = path.get(path.size()-1);
        for(int i = 0; i < containers_count; i++)
        {
            temp4 = last_path.get(i);
            if(temp4 == vol_wanted) return true;
        }
        return false;

    }

    public static double get_index(ArrayList<Integer> node, int containers_count)
    {
        // generating a random number for the key index using pow to be sure there's no duplicate
        double result_ind = 1;
        for(int i = 0; i < containers_count; i++)
        {
            result_ind = result_ind * Math.pow(3+i, node.get(i));
        }
        return result_ind;
    }
}



